module/builder/collection/builder.go

```diff
@@ -72,35 +73,41 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 
 	// first we construct a proposal in-memory, ensuring it is a valid extension
 	// of chain state -- this can be done in a read-only transaction
-	err := b.db.View(func(tx *badger.Txn) error {
+	err := b.db.View(func(btx *badger.Txn) error {
 
 		// STEP ONE: Load some things we need to do our work.
 		// TODO (ramtin): enable this again
 		// b.tracer.StartSpan(parentID, trace.COLBuildOnSetup)
 		// defer b.tracer.FinishSpan(parentID, trace.COLBuildOnSetup)
 
 		var parent flow.Header
-		err := operation.RetrieveHeader(parentID, &parent)(tx)
+		err := operation.RetrieveHeader(parentID, &parent)(btx)
 		if err != nil {
 			return fmt.Errorf("could not retrieve parent: %w", err)
 		}
 
 		// retrieve the height and ID of the latest finalized block ON THE MAIN CHAIN
 		// this is used as the reference point for transaction expiry
 		var refChainFinalizedHeight uint64
-		err = operation.RetrieveFinalizedHeight(&refChainFinalizedHeight)(tx)
+		err = operation.RetrieveFinalizedHeight(&refChainFinalizedHeight)(btx)
 		if err != nil {
 			return fmt.Errorf("could not retrieve main finalized height: %w", err)
 		}
 		var refChainFinalizedID flow.Identifier
-		err = operation.LookupBlockHeight(refChainFinalizedHeight, &refChainFinalizedID)(tx)
+		err = operation.LookupBlockHeight(refChainFinalizedHeight, &refChainFinalizedID)(btx)
 		if err != nil {
 			return fmt.Errorf("could not retrieve main finalized ID: %w", err)
 		}
+		// pre-compute the minimum possible reference block height for transactions
+		// included in this collection (actual reference height may be greater)
+		minPossibleRefHeight := refChainFinalizedHeight - uint64(flow.DefaultTransactionExpiry-b.config.ExpiryBuffer)
+		if minPossibleRefHeight > refChainFinalizedHeight {
```

Comment:
Do we need to check if this minPossibleRefHeight is higher than the root height?